### PR TITLE
Observe network state to upload any unsent envelopes

### PR DESCRIPTION
**Pull request by markushi**
## :scroll: Description
Adds a `ConnectionStatusProvider` mechanic, so we can observe changes to the connection state.
API wise this is already complete, only some more tests for `SendCachedEnvelopeFireAndForgetIntegration` are required.

## :bulb: Motivation and Context
Partially implements https://github.com/getsentry/sentry-java/issues/912

## :green_heart: How did you test it?
Added unit tests.

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

**Comment by romtsn** [comment]
Note that on Android we use [SendCachedEnvelopeIntegration](https://github.com/getsentry/sentry-java/blob/main/sentry-android-core/src/main/java/io/sentry/android/core/SendCachedEnvelopeIntegration.java), so the change has likely to be implemented for both `SendCachedEnvelopeIntegration` and `SendCachedEnvelopeFireAndForgetIntegration`.

The former one is android-only as we also check for startup crashes there and block in case there are any. The latter one is jvm/backend only, and has to be manually installed by our users. We could also try to combine those two into one, but I remember I didn't want to do that, because on pure jvm apps we don't really need it, see https://github.com/getsentry/sentry-java/pull/2277#issuecomment-1277655397

File: sentry/src/main/java/io/sentry/EnvelopeSender.java [hunk 1]
```diff
@@ -53,6 +57,14 @@ protected void processFile(final @NotNull File file, final @NotNull Hint hint) {
       return;
     }
 
+    if (envelopeCache.containsFile(file)) {
```
Comment (romtsn) [inline sentry/src/main/java/io/sentry/EnvelopeSender.java:60]: `m`: I think we should do the same in `OutboxSender` as we also install the second instance of `SendCachedEnvelopeIntegration` for sending cached files from the outbox folder

Comment (romtsn) [reply sentry/src/main/java/io/sentry/EnvelopeSender.java:60]: Also, a comment would be nice to mention why this is needed

File: sentry/src/main/java/io/sentry/transport/RateLimiter.java [hunk 2]
```diff
@@ -87,6 +87,11 @@ public RateLimiter(final @NotNull SentryOptions options) {
     return envelope;
   }
 
+  @Nullable
+  public Date getRateLimitedUntilFor(final @NotNull String itemType) {
```
Comment (romtsn) [inline sentry/src/main/java/io/sentry/transport/RateLimiter.java:91]: This is probably from a different PR?

File: sentry/src/main/java/io/sentry/cache/EnvelopeCache.java [hunk 3]
```diff
@@ -416,6 +425,11 @@ public void discard(final @NotNull SentryEnvelope envelope) {
     return ret.iterator();
   }
 
+  @Override
+  public synchronized boolean containsFile(@NotNull File file) {
```
Comment (romtsn) [inline sentry/src/main/java/io/sentry/cache/EnvelopeCache.java:429]: `m`: how about we rename this to something like `hasAlreadyProcessedFile(file)` to make it a bit less confusing on the calling side (where we actually processed cached files, but at the same time early-returning if it's already cached 😅 )

**Review by romtsn — APPROVED** [review]
A few minor things:
* Left some comments what I think make sense to change a bit
* A few tests missing for the `sentry` module
* A few `final` keywords missing

Otherwise looks great, I think this is a crucial improvement 🚀

**Review by adinauer — APPROVED** [review]
Gave this a quick look, LGTM.

File: sentry-android-core/src/main/java/io/sentry/android/core/SendCachedEnvelopeIntegration.java [hunk 4]
```diff
@@ -29,7 +39,8 @@ public SendCachedEnvelopeIntegration(
   @Override
   public void register(@NotNull IHub hub, @NotNull SentryOptions options) {
     Objects.requireNonNull(hub, "Hub is required");
-    final SentryAndroidOptions androidOptions =
+    this.hub = hub;
```
Comment (stefanosiano) [inline sentry-android-core/src/main/java/io/sentry/android/core/SendCachedEnvelopeIntegration.java:42]: you can merge the assignment
```suggestion
    this.hub = Objects.requireNonNull(hub, "Hub is required");
```

File: sentry-android-core/src/main/java/io/sentry/android/core/internal/util/AndroidConnectionStatusProvider.java [hunk 5]
```diff
@@ -7,40 +7,48 @@
...
-  public enum Status {
-    CONNECTED,
-    NOT_CONNECTED,
-    NO_PERMISSION,
-    UNKNOWN
-  }
+  private final @NotNull Context context;
+  private final @NotNull ILogger logger;
+  private final @NotNull BuildInfoProvider buildInfoProvider;
+  private final @NotNull Map<IConnectionStatusObserver, ConnectivityManager.NetworkCallback>
```
[the remark this answers is not in the document]
Comment (stefanosiano) [reply sentry-android-core/src/main/java/io/sentry/android/core/internal/util/AndroidConnectionStatusProvider.java:34]: not required from what i can see, but that definitely doesn't hurt, especially for future changes

File: sentry-android-core/src/main/java/io/sentry/android/core/internal/util/AndroidConnectionStatusProvider.java [hunk 6]
```diff
@@ -50,6 +58,50 @@ private ConnectivityChecker() {}
...
 
+  @Override
+  public @Nullable String getConnectionType() {
+    return getConnectionType(context, logger, buildInfoProvider);
+  }
+
+  @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+  @Override
+  public boolean addConnectionStatusObserver(@NotNull IConnectionStatusObserver observer) {
+    final ConnectivityManager.NetworkCallback callback =
```
Comment (stefanosiano) [inline sentry-android-core/src/main/java/io/sentry/android/core/internal/util/AndroidConnectionStatusProvider.java:69]: what about adding a guard here for min api?
like
```
if (buildInfoProvider.getSdkInfoVersion() < Build.VERSION_CODES.LOLLIPOP) {
    logger.log(SentryLevel.DEBUG, "addConnectionStatusObserver need Android LOLLIPOP+.");
    return false;
}
```

File: sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java [hunk 7]
```diff
@@ -54,20 +60,59 @@ public SendCachedEnvelopeFireAndForgetIntegration(
...
 
-  @SuppressWarnings("FutureReturnValueIgnored")
   @Override
-  public final void register(final @NotNull IHub hub, final @NotNull SentryOptions options) {
+  public void register(final @NotNull IHub hub, final @NotNull SentryOptions options) {
     Objects.requireNonNull(hub, "Hub is required");
     Objects.requireNonNull(options, "SentryOptions is required");
 
+    this.hub = hub;
+    this.options = options;
```
Comment (stefanosiano) [inline sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java:69]: you can merge assignments here

**Review by stefanosiano — APPROVED** [review]
looks great, just fix the comments left and we're good to go!

**Comment by dave-t-tran** [comment]
@romtsn I noticed these code changes aren't in https://github.com/getsentry/sentry-java/releases/tag/7.0.0-rc.1. Are these changes slated for a future 7.x release?

**Comment by romtsn** [comment]
@dave-t-tran they've been part of https://github.com/getsentry/sentry-java/releases/tag/7.0.0-beta.1 already, so the rc.1 also contains them

**Pull request merged**